### ISSUE TITLE
PickupCartHandler does not set customer when logged user picks up a cart

### ISSUE THREAD

**Issue by gorkalaucirica** [Sylius/ShopApiPlugin#259]
Given this request:
```
curl $API_DOMAIN/shop-api/carts/1234565121 -v \
    -H "Authorization: Bearer $SHOP_API_ACCESS_TOKEN" \
    -X POST \
    -H "Content-Type: application/json" \
     --data ' {"channel": "B2B"}'
```
I expect to see customer assigned to the picked cart, but customer remains null.

Just adding the following to `PickupCartHandler` fixes the problem:
```php
if($this->tokenStorage->getToken()->getUser() instanceof ShopUser) {
     $cart->setCustomer($this->tokenStorage->getToken()->getUser()->getCustomer());
}
```

Is this correct? If so I will send a PR

**Comment by GregoireHebert** [comment]
Is it not already done by `Sylius\Bundle\CoreBundle\EventListener\CartBlamerListener` ?

**Comment by gorkalaucirica** [comment]
CartBlameListener listens to this:

- sylius.user.security.implicit_login
- security.interactive_login

Cart in ShopApi is created using a token generated in client side and therefore there is no link between session and cart. Maybe a strategy CartContext has to be implemented for ShopApi

**Comment by GregoireHebert** [comment]
Oh, you are right ! But IMO it's an expected behavior.
In my point of view, sometimes, you need to use the shop API but your users might not be logged in yet.
So you'll use an "anonymous" customer user and still create a cart.

Then, in my use case, once the user is logged in, I force the same cart to get the users id assigned.
To keep control, I added an endpoint to do that 

```
    public function __invoke($token, Request $request): Response
    {
        $order = $this->orderRepository->findOneBy(['tokenValue' => $token]);
        if (null === $order) {
            throw new NotFoundHttpException("Cannot find any cart for token $token");
        }

        if (!$order->getCustomer()) {
            $order->setCustomer($this->tokenStorage->getToken()->getUser()->getCustomer());

            $memorisedOrder = $this->cartContext->getCart();
            if ($memorisedOrder && $order->getId() !== $memorisedOrder->getId()) {
                $this->manager->remove($memorisedOrder);
            }

            $this->manager->flush();
        }

        try {
            return $this->viewHandler->handle(
                View::create($this->cartQuery->getOneByToken($token), Response::HTTP_OK)
            );
        } catch (\InvalidArgumentException $exception) {
            throw new BadRequestHttpException($exception->getMessage());
        }
    }
```

**Comment by gorkalaucirica** [comment]
@GregoireHebert yes, but yours is another use case: "An anonymous user with an already picked cart logs in"
My use case is: "A logged user picks a cart".

CartContext does not work in PickupCart use case because is never used, PickupCartHandler just creates a cart and adds it to the repo.

**Comment by lchrusciel** [comment]
Sorry for a late response.

The clue of the problem is that we wanted to unify logic of cart processing as much as it is possible. So it doesn't matter if the customer is logged in or not until checkout complete. AFAIR the current solution is the simplest way to allow to log in to the customer whenever they want. It is the last moment for the customer to log in. Otherwise one has to provide an email. The feature you are asking for is partially implemented already in cart complete. It is covered by [tests here](https://github.com/Sylius/SyliusShopApiPlugin/blob/master/tests/Controller/CheckoutCompleteOrderApiTest.php#L131-L188). 

Anyway, if you want to implement this logic your proposal is not the best one because it will couple your handler with authorization, and you really don't want to do it :) I believe being stateless on handlers level is one of the most crucial changes that should be introduced to Sylius itself. 

You can always change PickupCommand and Handler so that you will pass a customer email inside of command and the Controller will be responsible for fetching the customer from token storage. 

A silver bullet for this logic would be to create yet another command that will claim the cart whenever the customer will log in, do it automatically if they are logged in during cart pickup or an email will be required for the last step of checkout but it will require some work and a lot of tests.

Anyway, the current behaviour is intentional so I believe this issue can be closed.

**Comment by gorkalaucirica** [comment]
@lchrusciel We developed a custom "Claim Cart" command that reads authentication header, associates the cart with the customer and recalculates the cart.

We need this in order to apply the customer related promotions (CustomerGroupRule).

We can send a PR, does this make sense to you?

**Comment by jdeveloper** [comment]
@lchrusciel what about when picking up a cart if there is an authenticated user assign its customer and in the complete action only require the customer if no one allready assigned?

I think it could be done in an easy form and you could still offer the flexebility of loggin when you want.

**Comment by lchrusciel** [comment]
@jdeveloper you are totally right, the case when the customer should be assigned to cart ASAP is for example for promotions (customer based) or taxation. It is relevant, and should be changed

**Comment by tophsic** [comment]
Related to #406, isn't it? I did not have time to search a better solution than #407 but as @lchrusciel and @mamazu said in this issue, the solution should not be done in each handler.

**Comment by mamazu** [comment]
Yes, if #406 is closed this will be solved as well (kinda). But yes. There are those two instances where assigning a customer would be needed.
a) on login of the customer the current cart gets attached to the user
b) on cart pickup. If the customer is logged in it will automatically assign the cart to the user.

Then completing the checkout will also become easier as there is either no customer logged in (guest checkout) or a logged in customer attached to the cart.

**Comment by lchrusciel** [comment]
Done in #488
